core/lib/spree/acts_as_tenant_model_extensions.rb

```diff
@@ -0,0 +1,163 @@
+      belongs_to tenant, scope, **valid_options
+
+      default_scope lambda {
+        if ActsAsTenant.should_require_tenant? && ActsAsTenant.current_tenant.nil? && !ActsAsTenant.unscoped?
+          raise ActsAsTenant::Errors::NoTenantSet
+        end
+
+        if ActsAsTenant.current_tenant
+          keys = [ActsAsTenant.current_tenant.send(pkey)].compact
+          keys.push(nil) if options[:has_global_records]
+
+          if options[:through]
+            query_criteria = { options[:through] => { fkey.to_sym => keys } }
+            query_criteria[polymorphic_type.to_sym] = ActsAsTenant.current_tenant.class.to_s if options[:polymorphic]
+            joins(options[:through]).where(query_criteria)
+          else
+            query_criteria = { fkey.to_sym => keys }
+            query_criteria[polymorphic_type.to_sym] = ActsAsTenant.current_tenant.class.to_s if options[:polymorphic]
+            where(query_criteria)
+          end
+        else
+          all
+        end
+      }
+
+      # Add the following validations to the receiving model:
+      # - new instances should have the tenant set
+      # - validate that associations belong to the tenant, currently only for belongs_to
+      #
+      before_validation proc { |m|
+        if ActsAsTenant.current_tenant
+          if options[:polymorphic]
+            m.send("#{fkey}=".to_sym, ActsAsTenant.current_tenant.class.to_s) if m.send(fkey.to_s).nil?
+            if m.send(polymorphic_type.to_s).nil?
+              m.send("#{polymorphic_type}=".to_sym, ActsAsTenant.current_tenant.class.to_s)
+            end
+          else
+            m.send "#{fkey}=".to_sym, ActsAsTenant.current_tenant.send(pkey)
+          end
+        end
+      }, on: :create
+
+      polymorphic_foreign_keys = reflect_on_all_associations(:belongs_to).select do |a|
+        a.options[:polymorphic]
+      end.map { |a| a.foreign_key }
+
+      reflect_on_all_associations(:belongs_to).each do |a|
+        ##### Custom #####
+        # The optional belongs_to association triggers errors.
+        # next if a.options[:optional] == true
+        ##### Custom #####
+
+        next if a == reflect_on_association(tenant) || polymorphic_foreign_keys.include?(a.foreign_key)
+
+        validates_each a.foreign_key.to_sym do |record, attr, value|
+          next if value.nil?
+          next unless record.will_save_change_to_attribute?(attr)
+
+          primary_key = if a.respond_to?(:active_record_primary_key)
+                          a.active_record_primary_key
+                        else
+                          a.primary_key
+                        end.to_sym
+          scope = a.scope || ->(relation) { relation }
+
+          ##### Custom #####
+          # The associated record is not saved in the database and this causes the error to be added to the model.
+          # next if record.association(a.name).target.send(primary_key) == value
+          ##### Custom #####
```

Comment:
This only solves a small portion of errors when the above `next if a.options[:optional] == true` is not used.